src/runtime/cuda.cpp

```diff
@@ -99,25 +100,37 @@ WEAK int halide_cuda_acquire_context(void *user_context, CUcontext *ctx, bool cr
     // not block execution on failure.
     halide_assert(user_context, ctx != NULL);
 
-    halide_assert(user_context, &thread_lock != NULL);
-    while (__sync_lock_test_and_set(&thread_lock, 1)) { }
-
     // If the context has not been initialized, initialize it now.
     halide_assert(user_context, &context != NULL);
-    if (context == NULL && create) {
-        CUresult error = create_cuda_context(user_context, &context);
-        if (error != CUDA_SUCCESS) {
-            __sync_lock_release(&thread_lock);
-            return error;
+
+    CUcontext local_val;
+    __atomic_load(&context, &local_val, __ATOMIC_ACQUIRE);
+    if (local_val == NULL) {
+        if (!create) {
+            *ctx = NULL;
+            return 0;
+        }
+
+        while (__sync_lock_test_and_set(&context_lock, 1)) { }
+
+        __atomic_load(&context, &local_val, __ATOMIC_ACQUIRE);
+        if (local_val == NULL) {
+            CUresult error = create_cuda_context(user_context, &local_val);
+            if (error != CUDA_SUCCESS) {
+                __sync_lock_release(&context_lock);
```

Comment:
Are there `__atomic` equivalents to `__sync_lock_release` and the other `__sync_ functions used here? (The GCC docs suggest that new code should always use the ‘__atomic’ builtins rather than the ‘__sync’ builtins.)

Reply:
There are, but I'd rather change all of our spinlock stuff at once.